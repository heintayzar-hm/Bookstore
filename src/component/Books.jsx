import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { addBookAction, removeBookAction } from '../redux/books/booksReducer';
import AddBook from './AddBook';
import Book from './Book';

// eslint-disable-next-line react/prefer-stateless-function
class Books extends React.Component {
  constructor(props) {
    super();
    this.props = props;
    this.addBook = this.addBook.bind(this);
    this.removeBook = this.removeBook.bind(this);
  }

  addBook(title, author) {
    const { dispatch } = this.props;
    dispatch(addBookAction({ title, author }));
  }

  removeBook(id) {
    const { dispatch } = this.props;
    dispatch(removeBookAction(id));
  }

  render() {
    const { books } = this.props;
    return (
      <>
        <ul>
          {books.map((book) => (
            <Book
              key={book.id}
              id={book.id}
              title={book.title}
              author={book.author}
              removeBook={this.removeBook}
            />
          ))}
        </ul>
        <AddBook addBook={this.addBook} />
      </>
    );
  }
}

const mapStateToProps = (state) => ({
  // specify which state data to provide to the component
  books: state.books,
});
Books.propTypes = {
  dispatch: PropTypes.func.isRequired,
  books: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      author: PropTypes.string,
    }),
  ).isRequired,
};
export default connect(mapStateToProps)(Books);
