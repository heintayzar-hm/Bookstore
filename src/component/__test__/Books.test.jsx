import { describe, expect, test } from 'vitest';
import { render } from '@testing-library/react';
import Books from '../Books';

describe('Books', () => {
  test('renders the initial state correctly', () => {
    // Arrange

    // Act
    const { container } = render(
      <Books />,
    );
    // Assert
    expect(container).toMatchSnapshot();
  });

  test('Have add new component', () => {
    // Arrange

    // Act
    const { getAllByText } = render(
      <Books />,
    );
    // Assert
    expect(getAllByText('Add New Book').length).toBe(1);
  });
});
