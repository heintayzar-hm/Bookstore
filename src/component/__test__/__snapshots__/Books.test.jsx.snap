// Vitest Snapshot v1

exports[`Books > renders the initial state correctly 1`] = `
<div>
  <ul />
  <article
    data-testid="form-container"
  >
    <h2>
      Add New Book
    </h2>
    <form>
      <input
        data-testid="book-title-input"
        name="title"
        placeholder="Book Title"
        required=""
        type="text"
        value=""
      />
      <input
        data-testid="author-title-input"
        name="author"
        placeholder="Author"
        required=""
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add Book
      </button>
    </form>
  </article>
</div>
`;
